### ISSUE TITLE
spawn vasmm68k_mot ENOENT

### ISSUE THREAD

**Issue by asimilon** [prb28/vscode-amiga-assembly#143]
The latest version of VSCode that just auto updated has started showing me this error when trying to launch the emulator.

> Cannot find bin/vasmm68k_mot : spawn bin/vasmm68k_mot ENOENT
> Source: Amiga Assembly (Extension)

This makes no sense to me because the vasm68k_mot file is in exactly the same place that it was last night when this working, but today VSCode updated itself to:

```
Version: 1.54.1
Commit: f30a9b73e8ffc278e71575118b6bf568f04587c8
Date: 2021-03-04T22:42:18.719Z (18 hrs ago)
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 20.3.0
```

I tried searching for some information about this ENOENT error but didn't find anything that I remotely understand.

**Comment by asimilon** [comment]
To solve my issue in the short term I have removed VSCode and installed the previous version from here: https://code.visualstudio.com/updates/v1_53

The download link is just below the title:

![image](https://user-images.githubusercontent.com/5970482/110151810-35e13f80-7de1-11eb-9e48-a4b9ce8f4459.png)

I can now build the executable and the emulator launches and runs it again.

I also switched updating for VSCode to manual.

**Comment by prb28** [comment]
Thanks for the workaround.
On windows it works well.
What is your os ?

**Comment by asimilon** [comment]
I'm using macOS 11.2

**Comment by prb28** [comment]
On osx 11.2 you must open each binary in bin folder with 'ctrl' key pressed, see "How to open an app that hasn’t been notarized or is from an unidentified developer ": https://support.apple.com/en-us/HT202491.
In the workspace file, please replace each binary path with it's absolute path.
It worked for me after.

**Comment by asimilon** [comment]
Hmmm, that is strange, I had done that already (well I stripped the quarantine bit which solves having to open every binary individually) and rolling back the VSCode version made everything work again.
For the workspace file is that the one named `*.code-workspace`?
Mine looks like this:
```
{
	"folders": [
		{
			"path": "."
		}
	],
	"settings": {
		"amiga-assembly.cstool": "bin/cstool",
		"amiga-assembly.vasm": {
			"enabled": true,
			"file": "bin/vasmm68k_mot",
			"options": [
				"-m68000",
				"-Fhunk",
				"-linedebug"
			]
		},
		"amiga-assembly.vlink": {
			"enabled": true,
			"file": "bin/vlink",
			"includes": "*.s",
			"excludes": "",
			"exefilename": "../uae/dh0/dfkt-hello",
			"options": [
				"-bamigahunk",
				"-Bstatic"
			]
		},
		"amiga-assembly.adfgenerator": {
			"ADFToolsParentDir": "bin",
			"sourceRootDir": "uae/dh0",
			"outputADFFile": "./build/disk.adf",
			"includes": "**/*",
			"excludes": "**/.*"
		},
		"editor.formatOnSave": true
	}
}
```
I presume I should be changing `"file": "bin/vasmm68k_mot"` to be the absolute path, rather than just `bin/vasmm68k_mot`?
I will try to update VSCode, apply those changes and be sure to manually open each binary and report back if I discover anything new.

**Comment by gljubojevic** [comment]
I can also confirm that for VSCode 1.55, MAC OS 11.2.3 setting tools path to absolute path in `.code-workspace` resolves issue.
```{
	"folders": [
		{
			"path": "."
		}
	],
	"settings": {
		"amiga-assembly.cstool": "bin/cstool",
		"amiga-assembly.vasm": {
			"file": "bin/vasmm68k_mot",
		},
		"amiga-assembly.vlink": {
			"file": "bin/vlink",
		},
	}
}
```

So instead e.g.
```
			"file": "bin/vasmm68k_mot",
```
put e.g. (where ever your project is located) **for all tools in bin**
```
			"file": "/Users/gljubojevic/qdev/AmigaDrawPoly/bin/vasmm68k_mot",
```

**For 3rd party tools used by extension...**

Another workaround for 3rd party tools on MAC OS instead of opening each tool with "ctrl" one by one and allowing it to run is to create script for tools download and extract.
Oddly enough this doesn't cause dialog saying that tool can't be verified and should be moved to bin.

Here is my version of script "get_osx_tools.sh"
```shell
# download osx tools for Amiga Assembly extension
echo "Downloading bin tools..."
echo "NOTE: No need to set osx unknown developer attributes when bin tools downloaded like this"
[ -d "bin" ] && rm -rf bin
mkdir -p bin
curl https://github.com/prb28/vscode-amiga-assembly/releases/download/0.21.1/osx.zip -L -o bin/osx.zip
unzip -d ./bin ./bin/osx.zip
rm ./bin/osx.zip
```

Might not be a bad idea to add script into: https://github.com/prb28/vscode-amiga-wks-example
to make it easier for others to use, I can do PR @prb28 if you like it.

**Comment by prb28** [comment]
@gljubojevic Thanks for the workaround. It's strange that the curl command does the trick.
It think it would be better to integrate it as a command in the extension, like "Download binaries in the workspace". Bartman adds his binaries in the extension directory not in the current workspace. I don't know what's better.

**Comment by gljubojevic** [comment]
Just a comment ....
I kind of like the idea that all tools are related to extension instead of workspace here is why
- no tools copy in each workspace, that removes burden of manual maintenance
- plugin can maintain appropriate tools e.g. automatic download, upgrade/update
- no need for tools path in workspace settings because plugin has paths of all tools

Issue on MAC os is that when you use finder and decompress tools OS probably adds attributes to forbid starting unsigned tools. I was investigating what is possible to make it easier instead running each and every file manually.
Other option instead of curl/unzip was:
```
xattr -w com.apple.quarantine "$(xaatr -p com.apple.quarantine adfcopy | sed 's/^.\{4\}/00c1/')" adfcopy
```
Found it here: https://transang.me/macos-open-file-from-unidentified-developers/
Making shell script to automate attribute setup I have discovered that simple curl/unzip does the trick.

Other interesting option would be not to install fs-uae with tools especially if it is available in system and has debug interface enabled not sure if latest version https://fs-uae.net/download#macosx has it.

**Comment by prb28** [comment]
> I kind of like the idea that all tools are related to extension instead of workspace here is why
>     * no tools copy in each workspace, that removes burden of manual maintenance

You can have it in one place and change the configuration's paths. But you're right it's more complex.

>     * plugin can maintain appropriate tools e.g. automatic download, upgrade/update

I agree.

>     * no need for tools path in workspace settings because plugin has paths of all tools

I agree.

The point of having it in the workspace or near is that I don't need to follow the releases of vasm/vlink. It's easier to update it without me having to build and push the new one.
If someone wants to stick to a version, it's easier too.

> Other interesting option would be not to install fs-uae with tools especially if it is available in system and has debug interface enabled not sure if latest version https://fs-uae.net/download#macosx has it.

Thanks for the macos tricks, I'll add it to the wiki.
No fs-uae and winuae main releases don't have the debugging feature.
They are waiting for the stability of the developments.

**Comment by asimilon** [comment]
> Other option instead of curl/unzip was:
> 
> ```
> xattr -w com.apple.quarantine "$(xaatr -p com.apple.quarantine adfcopy | sed 's/^.\{4\}/00c1/')" adfcopy
> ```

I found that `xattr -d com.apple.quarantine <filename>` was sufficient, iirc I used:

```
find . | xargs -d com.apple.quarantine
```

in the `bin` folder to apply to all the files.

**Comment by prb28** [comment]
Thanks.
Added a FAQ entry : https://github.com/prb28/vscode-amiga-assembly/wiki/FAQ
I'm working on a download feature integrated in the extension: will be added to the 0.22.

**Comment by prb28** [comment]
should be fixed by #152
